File: gateway/src/routes/testScenarios.ts
import type { FastifyInstance } from "fastify";
import { testScenarioStore, type TestScenarioRecord } from "../stores/contractStore.js";

export type TestScenario = TestScenarioRecord;

export async function testScenarioRoutes(app: FastifyInstance): Promise<void> {
	// GET /api/v1/test-scenarios - list all test scenarios
	app.get("/api/v1/test-scenarios", async (_request, reply) => {
		return reply.send(testScenarioStore.getAll());
	});

	// POST /api/v1/test-scenarios - add a new test scenario
	app.post("/api/v1/test-scenarios", async (request, reply) => {
		const body = request.body as Partial<TestScenario> | null;

		if (!body?.id || !body?.name || !body?.description) {
			return reply.status(400).send({
				error: "ValidationError",
				message: "Fields id, name, and description are required",
			});
		}

		const scenarios = testScenarioStore.getAll();
		if (scenarios.some((s) => s.id === body.id)) {
			return reply.status(409).send({
				error: "ConflictError",
				message: `Scenario with id '${body.id}' already exists`,
			});
		}

		const scenario: TestScenario = {
			id: body.id,
			name: body.name,
			description: body.description,
			inputSummary: body.inputSummary ?? "",
			expectations: body.expectations ?? [],
			requiredCapabilities: body.requiredCapabilities ?? [],
			requiresHumanReview: body.requiresHumanReview ?? false,
			prefersParallel: body.prefersParallel ?? false,
		};

		await testScenarioStore.add(scenario);
		return reply.status(201).send(scenario);
	});

	// DELETE /api/v1/test-scenarios/:id - remove a test scenario
	app.delete("/api/v1/test-scenarios/:id", async (request, reply) => {
		const { id } = request.params as { id: string };
		const removed = await testScenarioStore.remove(id);

		if (!removed) {
			return reply.status(404).send({
				error: "NotFound",
				message: `Scenario '${id}' not found`,
			});
		}
		return reply.send({ message: `Scenario '${id}' deleted` });
	});
}

File: gateway/python/routes/test_scenarios.py
"""Test scenarios routes."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..stores import test_scenario_store

router = APIRouter(prefix="/api/v1")


@router.get("/test-scenarios")
async def list_scenarios() -> list[dict[str, Any]]:
    return test_scenario_store.get_all()


@router.post("/test-scenarios")
async def add_scenario(request: Request) -> JSONResponse:
    body: dict[str, Any] = await request.json()

    if not body.get("id") or not body.get("name") or not body.get("description"):
        return JSONResponse(status_code=400, content={
            "error": "ValidationError",
            "message": "Fields id, name, and description are required",
        })

    existing = test_scenario_store.get_all()
    if any(s.get("id") == body["id"] for s in existing):
        return JSONResponse(status_code=409, content={
            "error": "ConflictError",
            "message": f"Scenario with id '{body['id']}' already exists",
        })

    scenario = {
        "id": body["id"],
        "name": body["name"],
        "description": body["description"],
        "inputSummary": body.get("inputSummary", ""),
        "expectations": body.get("expectations", []),
        "requiredCapabilities": body.get("requiredCapabilities", []),
        "requiresHumanReview": body.get("requiresHumanReview", False),
        "prefersParallel": body.get("prefersParallel", False),
    }
    await test_scenario_store.add(scenario)
    return JSONResponse(status_code=201, content=scenario)


@router.delete("/test-scenarios/{scenario_id}")
async def delete_scenario(scenario_id: str) -> JSONResponse:
    removed = await test_scenario_store.remove(scenario_id)
    if not removed:
        return JSONResponse(status_code=404, content={
            "error": "NotFound",
            "message": f"Scenario '{scenario_id}' not found",
        })
    return JSONResponse(content={"message": f"Scenario '{scenario_id}' deleted"})
